core/market/decentralized/src/negotiation/requestor.rs

```diff
@@ -219,9 +233,45 @@ impl RequestorBroker {
             .as_dao::<AgreementDao>()
             .save(agreement)
             .await
-            .map_err(|e| AgreementError::FailedSaveAgreement(proposal_id.clone(), e))?;
+            .map_err(|e| AgreementError::Save(proposal_id.clone(), e))?;
         Ok(id)
     }
+
+    /// Signs (not yet) Agreement self-created via `create_agreement`
+    /// and sends it to the Provider.
+    pub async fn confirm_agreement(
+        &self,
+        _id: Identity,
+        agreement_id: &AgreementId,
+    ) -> Result<(), AgreementError> {
+        let dao = self.db.as_dao::<AgreementDao>();
+        let agreement = match dao
+            .select(agreement_id, Utc::now().naive_utc())
+            .await
+            .map_err(|e| AgreementError::Get(agreement_id.clone(), e))?
+        {
+            None => return Err(AgreementError::NotFound(agreement_id.clone())),
+            Some(agreement) => agreement,
+        };
+
+        if agreement.state == AgreementState::Proposal {
+            self.api.propose_agreement(agreement).await?;
+            dao.update_state(agreement_id, AgreementState::Pending)
+                .await
+                .map_err(|e| AgreementError::Get(agreement_id.clone(), e))?;
+            return Ok(());
+        }
```

Comment:
Is it safe to call confirm_agreement form 2 concurrent threads?
(I think it won't hurt us, but I want be sure you verified it as well)

Reply:
moved to #430